File: backend/src/main/java/fr/sylvainjanet/tracker/tracking/application/service/CreateDailyRecordService.java
package fr.sylvainjanet.tracker.tracking.application.service;

import fr.sylvainjanet.tracker.tracking.application.port.in.exceptions.DailyRecordAlreadyExistsException;
import fr.sylvainjanet.tracker.tracking.application.port.in.usecase.CreateDailyRecordUseCase;
import fr.sylvainjanet.tracker.tracking.application.port.out.gateway.store.DailyRecordStore;
import fr.sylvainjanet.tracker.tracking.domain.DailyRecord;
import java.util.Objects;

public final class CreateDailyRecordService implements CreateDailyRecordUseCase {

    private final DailyRecordStore store;

    public CreateDailyRecordService(DailyRecordStore store) {
        this.store = Objects.requireNonNull(store, "store must not be null");
    }

    @Override
    public DailyRecord execute(DailyRecord command) throws DailyRecordAlreadyExistsException {
        Objects.requireNonNull(command, "command must not be null");

        return switch (store.create(command)) {
            case CREATED -> command;
            case ALREADY_EXISTS -> throw new DailyRecordAlreadyExistsException(command.date());
        };
    }
}

File: frontend/src/app/contexts/journal/application/service/create-daily-record.service.ts
import {
  type CreateDailyRecordResult,
  type CreateDailyRecordUseCase,
} from '../port/in/create-daily-record.use-case';
import { type DailyRecordGateway } from '../port/out/daily-record.gateway';
import type { DailyRecord } from '../../domain/daily-record';

export class CreateDailyRecordService implements CreateDailyRecordUseCase {
  constructor(private readonly gateway: DailyRecordGateway) {}

  async execute(command: DailyRecord): Promise<CreateDailyRecordResult> {
    const outcome = await this.gateway.create(command);

    if (outcome.kind === 'already-exists') {
      return {
        kind: 'already-exists',
      };
    }

    return {
      kind: 'created',
      resultData: outcome.outcomeData,
    };
  }
}
